feat: local authentication

nictownsend opened the pull request:
- introduce passport
- add a passport local strategy for authentication
- introduce Authentication interface to contain the multiple
auth checks (authenticate, checkAuth, logout)

Contributes to: strimzi/strimzi-ui#106

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

nictownsend commented:
Work to be done:

- [x] Further integration tests
- [x] MockAPI for `/login` purpose

matthew-chirgwin commented on linting/eslint.config.js:
```diff
@@ -12,7 +12,7 @@ const rulesets = [
 const customRules = {
   // prefer spaces (2) over tabs for indentation - tab width can be changed, space cannot
-  indent: ['error', 2],
+  indent: ['error', 2, { SwitchCase: 1 }],
```
What does this new config do (and should the comment be updated to match)?

nictownsend replied:
It was prettier and eslint disagreeing over `case` statement indenting; this just fixes that to state the amount of indenting expected.

matthew-chirgwin commented on server/client/client.feature:
```diff
@@ -5,44 +5,81 @@ Feature: client module
+    Scenario Outline: If assets can be served without authentication, the client module returns the appropriate <StatusCode> return code for a request of <Asset>
+        Given a server with a 'client' configuration
+        And There are files to serve
+        And authentication type 'none' is required
```
Being picky, behaviourally/grammatically, would it be better to have three seperate cases?

- And no authentication of the user is required
- And authentication via an identity provider 'type' of the user is required
- And username/password authentication of the user is required

nictownsend replied:
I thought I had different scenarios to cover that below?

matthew-chirgwin replied:
Sorry, I was more refering to the definition of the step - does a user care authentication type is none, or that they are/have been authenticated

nictownsend replied:
My opinion is that in these steps you're a "user" of the server. So you care around the config you have used to create the server.

matthew-chirgwin commented on server/client/client.feature:
```diff
@@ -5,44 +5,81 @@ Feature: client module
+            | /images/picture.svg | 200        |
+            | /doesnotexist.html  | 404        |
+            | /someroute          | 302        |
+            | /protected.html     | 200        |
```
This case (correct me if I am wrong) appears to be removed given the other test changes. The sceanrio here checks public assets can be retrieved without auth - but protected ones cannot

nictownsend replied:
I'm not sure I understand - how can protected files be blocked if there is no auth mechanism?

nictownsend replied:
I though the "no auth" case showed everything, but when "auth" is enabled, the protected routes required you to have logged in?

nictownsend replied:
Figured out what you mean (files that need auth cannot be retrieved when you're not authenticated) - refactoring now

matthew-chirgwin commented on server/client/controller.ts:
```diff
@@ -25,7 +25,6 @@ const publicFiles = [
   'images/*',
   'fonts/*',
   'favicon.ico',
-  'index.html',
```
Just to confirm, index now protected, as express will redirect to login if needed? What if auth is off?

nictownsend replied:
Yes, but if auth is off then `checkAuth` on protected routes is a noop

matthew-chirgwin commented on server/placeholderFunctionsToReplace.ts:
```diff
@@ -4,26 +4,4 @@
-};
-
-export { authFunction };
+export {};
```
Should this not be deleted if nothing uses it?

nictownsend replied:
Left as paradigm seemed useful for future, happy to remove.

matthew-chirgwin replied:
My preference would be that it is - as it has now been superceeded

matthew-chirgwin commented on server/security/router.ts:
```diff
@@ -0,0 +1,40 @@
+          res.send(200)
+        );
+        routerForModule.get(scram.login, (_req, res) =>
+          res.send('This will later be the login page')
```
comment to issue?

matthew-chirgwin commented on server/security/security.feature:
```diff
@@ -0,0 +1,62 @@
+  Scenario: SCRAM - authenticate valid credentials
+
+    Given a server with a 'security' configuration
+  # And authentication type 'scram' is required
```
Why are these disabled? Follow on work?

nictownsend replied:
Mistake, will add back :)

matthew-chirgwin commented on server/security/security.steps.ts:
```diff
@@ -0,0 +1,7 @@
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+import { Fusion } from 'jest-cucumber-fusion';
+import 'test_common/commonServerSteps';
```
this should not be needed - it should be provided for you

nictownsend replied:
I thought as much, but it doesn't - https://github.com/b-yond-infinite-network/jest-cucumber-fusion/blob/master/docs/ReusingStepDefinitions.md suggests the same

matthew-chirgwin replied:
I am very curious how any of the other tests work in this case. These steps should be preloaded for everyone here: https://github.com/strimzi/strimzi-ui/blob/ae571ee62875127d08716a42beb1ba9f58f79651/test_common/jest_cucumber_support/index.ts / https://github.com/strimzi/strimzi-ui/blob/74f9304bfd32ad7008c3f6eefd2bef8907a5d2cc/test_common/jest.common.config.js . If you want/need to interact with the world then it make sense to import the functions to be able to do that.

nictownsend replied:
Raised issue #146 - the existing tests work because they import the world and thus trigger the embedded given/when/then step definitions

matthew-chirgwin commented on server/security/strategy/strategyFactory.feature:
```diff
@@ -0,0 +1,13 @@
+    Then the returned strategy is of type '<expected>'
+    Examples:
+      | type    | expected |
+      | scram   | scram    |
```
oauth?

nictownsend replied:
To be added in the oauth PR :)

matthew-chirgwin commented on utils/dev_config/server.dev.config.js:
```diff
@@ -23,16 +23,20 @@ module.exports = {
   modules: {
     api: true,
-    client: false,
+    client: true,
```
When developing the UI (which this config is intended for), webpack hosts client files. As a result, we disable the module that would host the files here, to avoid confusion (why are my changes not present etc). What is this being enabled for?

nictownsend replied:
Artifact of testing, removing now.

pmuir reviewed (changes requested):
There seem to be some unrelated changes caught up in this PR. Please can you remove them and submit them separately.

nictownsend committed:
feat: local authentication

- introduce passport
- add a passport local strategy for authentication
- introduce Authentication interface to contain the multiple
auth checks (authenticate, checkAuth, logout)
- scram router module for authentication, logout, and auth check
- no op for no auth
- extend auth support to provide additional functions to all modules
for checking auth, logging out
- remove empty placeholder file

Contributes to: strimzi/strimzi-ui#106

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

nictownsend commented:
@pmuir sorted - they were whitespace changed from linting whole repo instead of staged files
